dll/win32/imm32/ctf.c

```diff
@@ -203,6 +228,328 @@ Imm32CheckAndApplyAppCompat(
     return pApphelpCheckIME(pszAppName);
 }
 
+/***********************************************************************
+ * TLS (Thread Local Storage) Related
+ */
+
+DWORD g_dwTLSIndex = -1;
+
+static VOID
+Imm32InitTLS(VOID)
+{
+    RtlEnterCriticalSection(&gcsImeDpi);
+
+    if (g_dwTLSIndex == -1)
+        g_dwTLSIndex = TlsAlloc();
+
+    RtlLeaveCriticalSection(&gcsImeDpi);
+}
+
+static IMMTLSDATA*
+Imm32AllocateTLS(VOID)
+{
+    IMMTLSDATA *pData;
+
+    if (g_dwTLSIndex == -1)
+        return NULL;
+
+    pData = (IMMTLSDATA*)TlsGetValue(g_dwTLSIndex);
+    if (pData)
+        return pData;
+
+    pData = (IMMTLSDATA*)ImmLocalAlloc(HEAP_ZERO_MEMORY, sizeof(IMMTLSDATA));
+    if (!pData)
+      return NULL;
+
+    if (!TlsSetValue(g_dwTLSIndex, pData))
+    {
+        ImmLocalFree(pData);
+        return NULL;
+    }
+
+    return pData;
+}
+
+static IMMTLSDATA*
+Imm32GetTLS(VOID)
+{
+    if (g_dwTLSIndex == -1)
+        return NULL;
+
+    return (IMMTLSDATA*)TlsGetValue(g_dwTLSIndex);
+}
+
+/* Get */
+static DWORD
+Imm32GetCoInitCountSkip(VOID)
+{
+    IMMTLSDATA *pData = Imm32GetTLS();
+    if (!pData)
+        return 0;
+    return pData->dwSkipCount;
+}
+
+/* Increment */
+static DWORD
+Imm32IncCoInitCountSkip(VOID)
+{
+    IMMTLSDATA *pData;
+    DWORD dwOldSkipCount;
+
+    pData = Imm32GetTLS();
+    if (!pData)
+        return 0;
+
+    dwOldSkipCount = pData->dwSkipCount;
+    if (pData->bDoCount)
+        pData->dwSkipCount = dwOldSkipCount + 1;
+
+    return dwOldSkipCount;
+}
+
+/* Decrement */
+static DWORD
+Imm32DecCoInitCountSkip(VOID)
+{
+    DWORD dwSkipCount;
+    IMMTLSDATA *pData;
+
+    pData = Imm32GetTLS();;
+    if (!pData)
+        return 0;
+
+    dwSkipCount = pData->dwSkipCount;
+    if (pData->bDoCount)
+    {
+        if (dwSkipCount)
+            pData->dwSkipCount = dwSkipCount - 1;
+    }
+
+    return dwSkipCount;
+}
+
+/***********************************************************************
+ * ISPY (I am not spy!)
```

Comment:
What's its purpose?

Reply:
COM-based Text Input Manager watches the `CoInitialize[Ex]`/`CoUninitialize` calls to get the timing of input activation/de-activation

Reply:
Implementation is still incomplete.